architecture/replication-layer.md

```diff
@@ -0,0 +1,114 @@
+
+Because this log is treated as serializable, it can be replayed to bring a node from a past state to its current state. This log also lets nodes that temporarily went offline to be "caught up" to the current state without needing to receive a copy of the existing data in the form of a snapshot.
+
+### Snapshots
+
+Each replica can be "snapshotted", which copies all of its data as of a specific timestamp (which is available because of [MVCC](storage-layer.html#mvcc)). This snapshot can be sent to other nodes during a rebalance event to expedite replication.
+
+After loading the snapshot, the node gets up to date by replaying all actions from the Raft group's log that have occurred since the snapshot was taken.
+
+### Leases
+
+A single node in the Raft group acts as the Lease Holder, which is the only node that can serve reads or propose writes to the Raft group leader (both actions are received as `BatchRequests` from [`DistSender`](distribution-layer.html#distsender)).
+
+When serving reads, Lease Holders bypass Raft; for the Lease Holders writes to have been committed in the first place, they must have already achieved consensus, so a second consensus on the same data is unnecessary. This has the benefit of not incurring networking round trips required by Raft and greatly increases the speed of reads (without sacrificing consistency).
+
+CockroachDB attempts to elect a Lease Holder who is also the Raft group leader, which can also optimize the speed of writes.
+
+If there is no Lease Holder, any node receiving a request will attempt to become to Lease Holder for the range. To prevent two nodes from acquiring the lease, the requester includes a copy of the last valid lease it had; if another node became the Lease Holder, its request is ignored.
+
+#### Co-location with Raft Leadership
+
+The range lease is completely separate from Raft leadership, and so without further efforts, Raft leadership and the Range lease might not be held by the same Replica. However, we can optimize query performance by making the same node both Raft leader and the Lease Holder; it reduces network round trips if the Lease Holder receiving the requests can simply propose the Raft commands to itself, rather than communicating them to another node.
+
+To achieve, this each lease renewal or transfer also attempts to collocate them. In practice, that means that the mismatch is rare and self-corrects quickly.
+
+#### Epoch-Based Leases (Table Data)
+
+To manage leases for table data, CockroachDB implements a notion of "epochs," which are defined as the period between a node joining a cluster and a node disconnecting from a cluster. When the node disconnects, the epoch is considered changed, and the node immediately loses all of it leases.
```

Comment:
`it leases` > `its leases`

But I don't think the node `immediately` loses leases when it disconnects. I think that happens only after 5 min.